libraries/bot-applicationinsights/src/test/java/com/microsoft/bot/applicationinsights/TelemetryWaterfallTests.java

```diff
@@ -0,0 +1,396 @@
+        ConversationState convoState = new ConversationState(new MemoryStorage());
+        TestAdapter adapter = new TestAdapter(TestAdapter.createConversationReference("ensureEndDialogCalled", "User1", "Bot"))
+            .use(new AutoSaveStateMiddleware(convoState));
+
+        StatePropertyAccessor<DialogState> dialogState = convoState.createProperty("dialogState");
+        DialogSet dialogs = new DialogSet(dialogState);
+        HashMap<String, Map<String, String>> saved_properties = new HashMap<>();
+        final Integer[] counter = {0};
+
+        // Set up the client to save all logged property names and associated properties (in "saved_properties").
+        BotTelemetryClient telemetryClient = new BotTelemetryClient() {
+            @Override
+            public void trackAvailability(String name, OffsetDateTime timeStamp, Duration duration, String runLocation, boolean success, String message, Map<String, String> properties, Map<String, Double> metrics) {
+
+            }
+
+            @Override
+            public void trackDependency(String dependencyTypeName, String target, String dependencyName, String data, OffsetDateTime startTime, Duration duration, String resultCode, boolean success) {
+
+            }
+
+            @Override
+            public void trackEvent(String eventName, Map<String, String> properties, Map<String, Double> metrics) {
+                StringBuilder sb = new StringBuilder(eventName).append("_").append(counter[0]++);
+                saved_properties.put(sb.toString(), properties);
+            }
+
+            @Override
+            public void trackException(Exception exception, Map<String, String> properties, Map<String, Double> metrics) {
+
+            }
+
+            @Override
+            public void trackTrace(String message, Severity severityLevel, Map<String, String> properties) {
+
+            }
+
+            @Override
+            public void trackDialogView(String dialogName, Map<String, String> properties, Map<String, Double> metrics) {
+
+            }
+
+            @Override
+            public void flush() {
+
+            }
+        };
+
+        MyWaterfallDialog waterfallDialog = new MyWaterfallDialog("test", newWaterfall());
+
+        dialogs.add(waterfallDialog);
+        dialogs.setTelemetryClient(telemetryClient);
+
+        new TestFlow(adapter, turnContext -> {
+            DialogContext dc = dialogs.createContext(turnContext).join();
+            dc.continueDialog().join();
+            if (!turnContext.getResponded()) {
+                dc.beginDialog("test", null).join();
+            }
+            return CompletableFuture.completedFuture(null);
+        })
+        .send("hello")
+        .assertReply("step1")
+        .send("hello")
+        .assertReply("step2")
+        .send("hello")
+        .assertReply("step3")
+        .send("hello")
+        .assertReply("step1")
+        .startTest()
+        .join();
+
+        Assert.assertEquals((Integer)7, counter[0]);
```

Comment:
Is it okay the `(Integer)7`?

Reply:
Otherwise it doesn't like it, IntelliJ suggests that casting to fix it.

Reply:
What if counter is `int[]` instead of `Integer`? I want to avoid the cast

Reply:
I will re review this, because counter had to be a very specific type because it was being used inside a sub clsas, but that was removed, so there might be a better way.

Reply:
I convert it to `int[]` as you suggested.
Also, the `final int[]` could not be replaced. [Here is an article](https://www.baeldung.com/java-lambda-effectively-final-local-variables) explaining why this happens.